torch/csrc/jit/node_hashing.cpp

```diff
@@ -19,27 +19,90 @@ bool tensorEqual(const at::Tensor& lhs, const at::Tensor& rhs) {
   return lhs.options().type_equal(rhs.options()) && lhs.equal(rhs);
 }
 
-bool tensorListEqual(
+bool typeListEqual(
+    const std::vector<TypePtr>& lhs,
+    const std::vector<TypePtr>& rhs) {
+  if (lhs.size() != rhs.size())
+    return false;
+  for (size_t i = 0; i < lhs.size(); ++i) {
+    if (*lhs[i] != *rhs[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename attribute_type> // int64_t, bool, double
+bool attributesEqual(attribute_type a1, attribute_type a2) {
+  return a1 == a2;
+}
+
+bool attributesEqual(const at::Tensor& a1, const at::Tensor& a2) {
+  return tensorEqual(a1, a2);
+}
+
+bool ivaluesEqual(const IValue& a1, const IValue& a2);
+
+bool attributesEqual(
     const std::vector<at::Tensor>& lhs,
     const std::vector<at::Tensor>& rhs) {
   if (lhs.size() != rhs.size())
     return false;
   return std::equal(lhs.begin(), lhs.end(), rhs.begin(), tensorEqual);
 }
 
-bool typeListEqual(
-    const std::vector<TypePtr>& lhs,
-    const std::vector<TypePtr>& rhs) {
-  if (lhs.size() != rhs.size())
+bool attributesEqual(at::ArrayRef<IValue> a1, at::ArrayRef<IValue> a2) {
+  if (a1.size() != a2.size()) {
     return false;
-  for (size_t i = 0; i < lhs.size(); ++i) {
-    if (*lhs[i] != *rhs[i]) {
+  }
+  for (size_t i = 0; i < a1.size(); ++i) {
+    if (!ivaluesEqual(a1[i], a2[i])) {
       return false;
     }
   }
   return true;
 }
 
+bool attributesEqual(const IValue& a1, const IValue& a2) {
+  return ivaluesEqual(a1, a2);
+}
+
+// this is not a general-purpose comparison of IValues, it only covers the
+// ivalues that are allowed as attributes, and it does not check type
+// equivalence of containers.
+bool ivaluesEqual(const IValue& a1, const IValue& a2) {
+  if (a1.tagKind() != a2.tagKind()) {
+    return false;
+  }
+  if (a1.isInt()) {
+    return a1.toInt() == a2.toInt();
+  }
+  if (a1.isBool()) {
+    return a1.toBool() == a2.toBool();
+  }
+  if (a1.isDouble()) {
+    return a1.toDouble() == a2.toDouble();
+  }
+  if (a1.isTensor()) {
+    return attributesEqual(a1.toTensor(), a2.toTensor());
+  }
+  if (a1.isNone()) {
+    return true;
+  }
+  if (a1.isString()) {
+    return attributesEqual(a1.toStringRef(), a2.toStringRef());
```

Comment:
is this not an infinite loop? there's no overload of `attributesEqual` for `std::string` as far as I can tell and the default just calls `ivaluesEqual`

Reply:
i think it resolves correctly because it doesn't infinite loop, but agreed it is confusing. will change.